### PR TITLE
bazel: add more detail to `dev test` cobra msg

### LINKED ISSUE #82411: bazel: add more detail to `./dev test --help`
- Currently, the cobra message does not describe the formatting or potential arguments for `--test-args` and `-stress-args`. I think the following could be specified like:
   - `stress-args`: additional arguments to pass to stress, prefixed with `-`. E.g. `--stress-args='-maxtime=30s'`. See all stress args here https://github.com/cockroachdb/stress/blob/master/main.go
   - `test-args` : additional arguments to pass to the go test binary, prefixed with `-test.{arg}`. For example: `--test-args='-test.timeout 1s'`. Learn about go test args here [??]. (I'm not sure if the msg should point to a CRDB specific list of go test args)

- Currently, the cobra message does not note that the user can pass bazel specific args behind `--`. Perhaps in the flag list, we could add something like:
    - `--`: add additional arguments to bazel, prefixed with `--`, e.g.` -- --sandbox_debug`. Note: the `dev` maintainers would like all commonly used bazel args to get passed directly to dev. Please file an issue if you're commonly reaching for the bazel flag!

- Currently, its unclear which flags operate on the dev level or go test level. It could be nice to specify which dev flags are merely shortcuts to go test, (e.g. `-verbose`)

Jira issue: CRDB-16361

Epic CRDB-15061

### PR DESCRIPTION
**Pull request by msbutler**
Fixes #82411

Release note: None

### CODE REVIEW

**Comment by msbutler** [comment]
If I rerun `./dev build short`, the cobra message does not reflect my changes. What's the easiest way to rebuild the dev package to inspect my changes?

File: pkg/cmd/dev/test.go [hunk 1]
```diff
@@ -73,16 +74,21 @@ pkg/kv/kvserver:kvserver_test) instead.`,
...
 	// visible.
-	testCmd.Flags().BoolP(vFlag, "v", false, "show testing process output")
+	testCmd.Flags().BoolP(vFlag, "v", false, "show testing process output ("+
+		"equivalent to passing '--testargs='test.v')")
 	testCmd.Flags().Int(countFlag, 1, "run test the given number of times")
 	testCmd.Flags().BoolP(showLogsFlag, "", false, "show crdb logs in-line")
 	testCmd.Flags().Bool(stressFlag, false, "run tests under stress")
-	testCmd.Flags().String(stressArgsFlag, "", "additional arguments to pass to stress")
+	testCmd.Flags().String(stressArgsFlag, "", "additional arguments to pass to stress.\n"+
+		"Prefix each arg with '-'; e.g. '--stress-args='-maxtime=30s'.\n"+
```
Comment (rickystewart) [inline pkg/cmd/dev/test.go:83]: Nit: An example with multiple arguments would be really helpful. e.g. `--stress-args='-maxtime=30s -maxruns=100'`

**Review by rickystewart — CHANGES_REQUESTED** [review]
Nit: `detail` in the commit message, not `detial`.

File: pkg/cmd/dev/test.go [hunk 2]
```diff
@@ -73,16 +74,21 @@ pkg/kv/kvserver:kvserver_test) instead.`,
 	// or only on failures. `--show-logs` by contrast is a flag for the process
 	// under test, controlling whether the process-internal logs are made
 	// visible.
-	testCmd.Flags().BoolP(vFlag, "v", false, "show testing process output")
+	testCmd.Flags().BoolP(vFlag, "v", false, "show testing process output ("+
+		"equivalent to passing '--testargs='test.v')")
```
[the remark this answers is not in the document]
Comment (msbutler) [reply pkg/cmd/dev/test.go:78]: Interesting! I was under the impression that `./dev test -v` would pass the `verbose` arg directly to the `go test` build but I guess not?

[the remark this answers is not in the document]
Comment (rickystewart) [reply pkg/cmd/dev/test.go:78]: It does do that. It also changes the `--test_output` argument for `bazel`.

File: pkg/cmd/dev/dev.go [hunk 3]
```diff
@@ -132,7 +132,12 @@ Typical usage:
 
 	// Add all the shared flags.
 	var debugVar bool
+	var bazelcmd string
 	ret.cli.PersistentFlags().BoolVar(&debugVar, "debug", false, "enable debug logging for dev")
+	ret.cli.PersistentFlags().StringVar(&bazelcmd, "", "", "pass additional arguments to bazel. "+
+		"Prefix each arg with '--', e.g. -- --sandbox_debug.\n"+
```
[the remark this answers is not in the document]
Comment (msbutler) [reply pkg/cmd/dev/dev.go]: done.

File: pkg/cmd/dev/dev.go [hunk 4]
```diff
@@ -132,7 +132,12 @@ Typical usage:
...
+		"Note: the dev maintainers would like all commonly used bazel args to get passed directly to"+
```
[the remark this answers is not in the document]
Comment (msbutler) [reply pkg/cmd/dev/dev.go]: Good to know! Deleted.

File: pkg/cmd/dev/test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (msbutler) [reply pkg/cmd/dev/test.go]: oh good to know. Changed the flag description to reflect this.

File: pkg/cmd/dev/test.go (hunk 1 above)
Comment (msbutler) [reply pkg/cmd/dev/test.go]: Done.

**Review by rickystewart — APPROVED** [review]

**Review by irfansharif — APPROVED** [review]
LGTM, appended a cosmetic commit to unify how the help text looks across `dev test -h` and `dev bench -h` -- feel free to drop it.

File: pkg/cmd/dev/test.go [hunk 5]
```diff
@@ -83,21 +77,16 @@ pkg/kv/kvserver:kvserver_test) instead.`,
...
 	// under test, controlling whether the process-internal logs are made
 	// visible.
-	testCmd.Flags().BoolP(vFlag, "v", false, "show testing process output ("+
-		"equivalent to passing '--testargs=test.v -- --test_output all')")
+	testCmd.Flags().BoolP(vFlag, "v", false, "show testing process output")
 	testCmd.Flags().Int(countFlag, 1, "run test the given number of times")
 	testCmd.Flags().BoolP(showLogsFlag, "", false, "show crdb logs in-line")
 	testCmd.Flags().Bool(stressFlag, false, "run tests under stress")
-	testCmd.Flags().String(stressArgsFlag, "", "additional arguments to pass to stress.\n"+
-		"Prefix each arg with '-'; e.g. '--stress-args='-maxtime=30s -maxruns=100'.\n"+
```
Comment (msbutler) [inline pkg/cmd/dev/test.go]: @irfansharif your commit reverts the additional language I added to each flag description. Was this intentional?

File: pkg/cmd/dev/dev.go [hunk 6]
```diff
@@ -132,10 +135,7 @@ Typical usage:
 
 	// Add all the shared flags.
 	var debugVar bool
-	var bazelcmd string
 	ret.cli.PersistentFlags().BoolVar(&debugVar, "debug", false, "enable debug logging for dev")
-	ret.cli.PersistentFlags().StringVar(&bazelcmd, "", "",
```
Comment (msbutler) [inline pkg/cmd/dev/dev.go]: @irfansharif ditto. was this intentional?

File: pkg/cmd/dev/test.go (hunk 5 above)
Comment (irfansharif) [reply pkg/cmd/dev/test.go]: Yup, this was moved to the top level help text with examples. Before:

```
      --stress-args string   additional arguments to pass to stress.
                             Prefix each arg with '-'; e.g. '--stress-args='-maxtime=30s -maxruns=100'.
                             See all stress args at https://github.com/cockroachdb/stress/blob/master/main.go
      --test-args string     additional arguments to pass to the go test binary.
                             Pass each arg with '-test.{arg}'; for example: --test-args='-test.timeout 1s'.
                             Learn about go test args by running 'go help test'.
```

Now:

```
Examples:

        dev test
        dev test pkg/kv/kvserver --filter=TestReplicaGC* -v --timeout=1m
        dev test pkg/... -v
        dev test pkg/spanconfig/... pkg/ccl/spanconfigccl/...                # test multiple packages recursively
        dev test --race --stress ...                                         # run a test under race and stress
        dev test pkg/spanconfig/... --test-args '-test.trace=trace.out'      # pass arguments to the go test binary (see 'go help testflag')
        dev test pkg/spanconfig --stress --stress-args '-maxruns 1000 -p 4'  # pass arguments to github.com/cockroachdb/stress
        dev test --stress --timeout=1m --test-args='-test.timeout 5s'        # stress for 1m until a test runs longer than 5s
        dev test pkg/server -f=TestSpanStatsResponse -v --count=5 --vmodule='raft=1'
```

File: pkg/cmd/dev/dev.go (hunk 6 above)
Comment (irfansharif) [reply pkg/cmd/dev/dev.go]: Yup, they're moved to the top level help text. Before it looked like this which I found confusing:

```
Flags:
      -- string   pass additional arguments directly to bazel; e.g. '-- --sandbox_debug'.
      --debug     enable debug logging for dev
  -h, --help      help for dev
  -v, --version   version for dev
```

Now it looks like this:

```
Typical usage:
...
    dev build short -- --verbose_failures --profile=prof.gz
                Pass additional arguments directly to bazel.
```

File: pkg/cmd/dev/test.go (hunk 5 above)
Comment (msbutler) [reply pkg/cmd/dev/test.go]: works for me! Is it cool if i move the description above each example, as I had originally? I think its best to keep each line under 70 chars. The help msg currently looks like this in a small terminal window:
image

Comment (irfansharif) [reply pkg/cmd/dev/test.go]: Works for me. An alternative is perhaps making this look like like `dev -h`, but I'm getting too bikesheddy.

Comment (msbutler) [reply pkg/cmd/dev/test.go]: Good idea. Merged your pr. Let me know if it looks good!

**Review by irfansharif — APPROVED** [review]
Thanks Michael, this will really help!

File: pkg/cmd/dev/test.go [hunk 7]
```diff
@@ -52,12 +52,29 @@ This command takes a list of packages and tests all of them. It's also
...
 pkg/kv/kvserver:kvserver_test) instead.`,
 		Example: `
-	dev test
-	dev test pkg/kv/kvserver --filter=TestReplicaGC* -v --timeout=1m
-	dev test pkg/server -f=TestSpanStatsResponse -v --count=5 --vmodule='raft=1'
-	dev test pkg/spanconfig/... pkg/ccl/spanconfigccl/...
-	dev test pkg/... -v
-	dev test --stress --race ...`,
+    dev test pkg/kv/kvserver --filter=TestReplicaGC*
+        run the tests that match with 'TestReplicaGC*' in the kvserver package.
```
[the remark this answers is not in the document]
Comment (msbutler) [reply pkg/cmd/dev/test.go]: Done.

**Comment by msbutler** [comment]
bors r=irfansharif, rickystewart

**Pull request merged**